prime-router/src/main/kotlin/fhirengine/engine/FHIRTranslator.kt

```diff
@@ -68,36 +71,76 @@ class FHIRTranslator(
                 val receiver = settings.findReceiver(message.receiverFullName)
                     ?: throw RuntimeException("Receiver with name ${message.receiverFullName} was not found")
                 actionHistory.trackActionReceiverInfo(receiver.organizationName, receiver.name)
-                val bodyBytes = getByteArrayFromBundle(receiver, bundle)
+
+                var nextAction = Event.EventAction.BATCH
+                var externalName = ""
+                var queueMessage: ReportEventQueueMessage? = null
+                val bodyBytes = if (message.topic.isSendOriginal) {
+                    nextAction = Event.EventAction.SEND
+                    val originalMessage = getOriginalMessage(message.reportId, WorkflowEngine())
+                    externalName = originalMessage.externalName
+                    queueMessage = ReportEventQueueMessage(nextAction, false, message.reportId, "")
+                    BlobAccess.downloadBlobAsByteArray(originalMessage.bodyUrl)
+                } else {
+                    getByteArrayFromBundle(receiver, bundle)
+                }
 
                 // get a Report from the message
                 val (report, event, blobInfo) = Report.generateReportAndUploadBlob(
-                    Event.EventAction.BATCH,
+                    nextAction,
                     bodyBytes,
                     listOf(message.reportId),
                     receiver,
                     this.metadata,
                     actionHistory,
                     topic = message.topic,
+                    externalName
                 )
 
+                if (queueMessage != null) {
+                    queueMessage = queueMessage.copy(reportId = report.id)
+                }
+
                 return listOf(
                     FHIREngineRunResult(
                         event,
                         report,
                         blobInfo.blobUrl,
-                        null
+                        queueMessage
                     )
                 )
             }
             else -> {
                 throw RuntimeException(
-                    "Message was not a FhirConvert and cannot be processed: $message"
+                    "Message was not a FhirTranslateQueueMessage and cannot be processed by FHIRTranslator: $message"
                 )
             }
         }
     }
 
+    /**
+     * Takes a [reportId] and returns the content of the original message as a ByteArray
+     */
+    internal fun getOriginalMessage(
```

Comment:
rename to getOriginalReport since you changed what this returns now. Also, @mkalish , do you think these two helpers would fit better in WorkflowEngine? Or any other ideas?

Reply:
I will create a follow-up ticket that includes both of these items and link it here. We can move forward without making this change.